snakemake/common/__init__.py

```diff
@@ -268,3 +271,17 @@ def get_input_function_aux_params(func, candidate_params):
     func_params = get_function_params(func)
 
     return {k: v for k, v in candidate_params.items() if k in func_params}
+
+
+_pool = concurrent.futures.ThreadPoolExecutor()
+
+
+@contextlib.asynccontextmanager
+async def async_lock(_lock: threading.Lock):
+    """Use a threaded lock form threading.Lock in an async context"""
+    loop = asyncio.get_event_loop()
+    await loop.run_in_executor(_pool, _lock.acquire)
+    try:
+        yield  # the lock is held
+    finally:
+        _lock.release()
```

Comment:
Instead of this, I am pretty sure we should just asyncio.Lock: https://docs.python.org/3/library/asyncio-sync.html#asyncio.Lock

Reply:
Can you modify the PR accordingly?

Reply:
The issue here is that `asyncio.Lock` is not thread safe. So for portions of the program coming at the resources from different threads, the asyncio locks will seize up and freeze the program (at least based on my tests). So long as there's mixed threading and async, it will need to use this function

Reply:
(I meant to leave an attribution to the stackoverflow source for this... I'll do that now)